### PR TITLE
Add support for GCP and S3 in `deploy`

### PR DESCRIPTION
**Pull request by swtwsk**
Depend on #3

---
Keep in mind:
- [X] Documentation updates
- [X] [Changelog](CHANGELOG.md) updates

### CODE REVIEW

File: data_pipelines_cli/cli_commands/deploy.py [hunk 1]
```diff
@@ -54,30 +57,29 @@ def _datahub_ingest():
...
-            "rsync",
-            "-d",
-            "-r",
-            str(build_dag_path),
-            f"{gs_bucket}/dags/pipeline-example",
-        ]
+def _sync_bucket(blob_address: str, remote_fs: AbstractFileSystem) -> None:
+    echo_info("Syncing Bucket")
+    blob_address_path = pathlib.PurePath(blob_address).joinpath(
+        "dags", "pipeline-example"
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/cli_commands/deploy.py]: Changed for `blob_protocol` field in `~/.dp.yml`

File: data_pipelines_cli/cli_commands/deploy.py [hunk 2]
```diff
@@ -54,30 +57,29 @@ def _datahub_ingest():
...
+    )
+    synchronize_local_remote_directory(
+        remote_fs,
+        BUILD_DIR.joinpath("dag"),
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/cli_commands/deploy.py]: `DeployCommand` class created

File: data_pipelines_cli/filesystem_utils.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            return {
+                "key": self.key,
+                "secret": self.secret,
+            }
+
+        echo_error(
+            "S3 expects either `--blob-token` or both `--blob-key` and "
+            "`--blob-secret` arguments to be provided"
+        )
+        sys.exit(1)
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/filesystem_utils.py]: Right now `dp deploy` expects path to JSON or YAML file with kwargs for specific fs.

File: data_pipelines_cli/filesystem_utils.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        filesystem_providers["gcp"] = lambda pk: gcsfs.GCSFileSystem(**pk.gcp())
+
+    try:
+        import s3fs
+    except ModuleNotFoundError:
+        pass
+    else:
+        filesystem_providers["s3"] = lambda pk: s3fs.S3FileSystem(**pk.s3())
+
+    return filesystem_providers
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/filesystem_utils.py]:
```py
class FilesystemProvider(Enum):
    GCP = "gcp"

def _get_gcs_fs(**kwargs) -> AbstractFileSystem:
    import gcsfs

    return gcsfs.GCSFileSystem(**kwargs)

FILESYSTEM_PROVIDERS = {
    FilesystemProvider.GCP.value: _get_gcs_fs,
}
```
is doing "lazy import evaluation" now

File: data_pipelines_cli/filesystem_utils.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            choices = " | ".join(FILESYSTEM_PROVIDERS.keys())
+            echo_error(
+                "More than one filesystem provider available. Please choose "
+                "one and run command with `--blob-provider` argument. "
+                f"Providers to choose: {choices}"
+            )
+            sys.exit(1)
+        provider = next(iter(FILESYSTEM_PROVIDERS.keys()))
+
+    return FILESYSTEM_PROVIDERS.get(provider)(provider_kwargs)
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/filesystem_utils.py]: removed

File: data_pipelines_cli/filesystem_utils.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    # TODO: Is it "lazy" (checking what to update) or not?
+    local_directory = local_fs.find(local_path)
+    local_directory_suffixes = set()
+    for local_file in local_directory:
+        local_file_suffix = local_file[len(local_path) :]
+        local_directory_suffixes.add(local_file_suffix)
+        remote_fs.put_file(local_file, remote_path + local_file_suffix)
+
+    if not delete:
+        return
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/filesystem_utils.py]: To be honest, I like that kind of "short circuiting" and I'm accustomed to using it. However, I have changed it, as requested :)

File: data_pipelines_cli/filesystem_utils.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    # Remove every file from remote that's not local
+    remote_directory = remote_fs.find(remote_path)
+    for remote_file in remote_directory:
+        remote_file_suffix = remote_file[len(remote_path) :]
+        if remote_file_suffix not in local_directory_suffixes:
+            remote_fs.rm(remote_file)
```
[the remark this answers is not in the document]
Comment (swtwsk) [reply data_pipelines_cli/filesystem_utils.py]: done

**Review by p-pekala — APPROVED** [review]

**Pull request merged**